screeps-game-api/src/memory.rs

```diff
@@ -42,46 +97,60 @@ impl MemoryReference {
         MemoryReference(reference)
     }
 
-    pub fn bool(&self, path: &str) -> bool {
-        js_unwrap!(Boolean(@{self.as_ref()}[@{path}]))
+    pub fn bool(&self, key: &str) -> bool {
+        js_unwrap!(Boolean(@{self.as_ref()}[@{key}]))
     }
 
-    pub fn num(&self, path: &str) -> Option<f64> {
+    pub fn path_bool(&self, path: &str) -> bool {
+        js_unwrap!(Boolean(_.get(@{self.as_ref()}, @{path})))
+    }
+
+    pub fn f64(&self, key: &str) -> Result<Option<f64>, ConversionError> {
         (js! {
-            return (@{self.as_ref()})[@{path}];
+            return (@{self.as_ref()})[@{key}];
         }).try_into()
-            .map(Some)
-            .unwrap_or_default()
     }
 
-    pub fn int(&self, path: &str) -> Option<i32> {
+    pub fn path_f64(&self, path: &str) -> Result<Option<f64>, ConversionError> {
         (js! {
-            return (@{self.as_ref()})[@{path}];
+            return _.get(@{self.as_ref()}, @{path});
         }).try_into()
-            .map(Some)
-            .unwrap_or_default()
     }
 
-    pub fn string(&self, path: &str) -> Option<String> {
+    pub fn i32(&self, key: &str) -> Result<Option<i32>, ConversionError> {
         (js! {
-            return (@{self.as_ref()})[@{path}];
+            return (@{self.as_ref()})[@{key}];
         }).try_into()
-            .map(Some)
-            .unwrap_or_default()
     }
 
-    pub fn dict(&self, path: &str) -> Option<MemoryReference> {
+    pub fn path_i32(&self, path: &str) -> Result<Option<i32>, ConversionError> {
         (js! {
-            var v = (@{self.as_ref()})[@{path}];
-            if (_.isArray(v)) {
-                return null;
-            } else {
-                return v || null;
-            }
+            return _.get(@{self.as_ref()}, @{path});
+        }).try_into()
+    }
+
+    pub fn string(&self, key: &str) -> Result<Option<String>, ConversionError> {
+        (js! {
+            return (@{self.as_ref()})[@{key}];
+        }).try_into()
+    }
+
+    pub fn path_string(&self, path: &str) -> Result<Option<String>, ConversionError> {
+        (js! {
+            return _.get(@{self.as_ref()}, @{path});
+        }).try_into()
+    }
+
+    pub fn dict(&self, key: &str) -> Result<Option<MemoryReference>, ConversionError> {
+        (js! {
+            return _.get(@{self.as_ref()}, @{key});
```

Comment:
Without the array check this will incorrectly succeed on arrays since they are `Reference`s.

Don't need the whole `v || null` thing (not sure why I felt the need to replace `undefined` with `null`), but the `if (_.isArray(v))` check is necessary to prevent incorrectly returning a `MemoryReference` wrapping an array.

Reply:
Actually, that should be taken care of by the defn of `TryFrom<Value> for MemoryReference` given later since I'm not wrapping the `MemoryReference` around the `Reference` myself.